pkg/neonvm/cpuscaling/sysfs.go

```diff
@@ -0,0 +1,195 @@
+package cpuscaling
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+)
+
+// CPU directory path
+const cpuPath = "/sys/devices/system/cpu/"
+
+type CPUSysFsStateScaler struct{}
+
+func (c *CPUSysFsStateScaler) EnsureOnlineCPUs(targetCount int) error {
+	cpus, err := getAllCPUs()
+	if err != nil {
+		return err
+	}
+
+	onlineCount, err := c.ActiveCPUsCount()
+	if err != nil {
+		return err
+	}
+
+	if onlineCount < uint32(targetCount) {
+		for _, cpu := range cpus {
+			if cpu == 0 {
+				// Skip CPU 0 as it is always online and can't be toggled
+				continue
+			}
+
+			online, err := isCPUOnline(cpu)
+			if err != nil {
+				return err
+			}
+
+			if !online {
+				// Mark CPU as online
+				err := setCPUOnline(cpu, true)
+				if err != nil {
+					return err
+				}
+				onlineCount++
+			}
+
+			// Stop when we reach the target count
+			if onlineCount == uint32(targetCount) {
+				break
+			}
+		}
+	} else if onlineCount > uint32(targetCount) {
+		// Remove CPUs if there are more than X online
+		for i := len(cpus) - 1; i >= 0; i-- {
+			cpu := cpus[i]
+			if cpu == 0 {
+				// Skip CPU 0 as it cannot be taken offline
+				continue
+			}
+
+			online, err := isCPUOnline(cpu)
+			if err != nil {
+				return err
+			}
+
+			if online {
+				// Mark CPU as offline
+				err := setCPUOnline(cpu, false)
+				if err != nil {
+					return err
+				}
+				onlineCount--
+			}
+
+			// Stop when we reach the target count
+			if onlineCount == uint32(targetCount) {
+				break
+			}
+		}
+	}
+
+	if onlineCount != uint32(targetCount) {
+		return fmt.Errorf("failed to ensure %d CPUs are online, current online CPUs: %d", targetCount, onlineCount)
+	}
+
+	return nil
+}
+
+// ActiveCPUsCount() returns the count of online CPUs.
+func (c *CPUSysFsStateScaler) ActiveCPUsCount() (uint32, error) {
+	cpus, err := getAllCPUs()
+	if err != nil {
+		return 0, err
+	}
+
+	var onlineCount uint32
+	for _, cpu := range cpus {
+		online, err := isCPUOnline(cpu)
+		if err != nil {
+			return 0, err
+		}
+		if online {
+			onlineCount++
+		}
+	}
+
+	return onlineCount, nil
+}
+
+// GetTotalCPUsCount returns the total number of CPUs (online + offline).
+func (c *CPUSysFsStateScaler) GetTotalCPUsCount() (uint32, error) {
+	cpus, err := getAllCPUs()
+	if err != nil {
+		return 0, err
+	}
+
+	return uint32(len(cpus)), nil
+}
+
+// Helper functions
+
+// getAllCPUs returns a list of all CPUs that are physically present in the system.
+func getAllCPUs() ([]int, error) {
+	data, err := os.ReadFile(filepath.Join(cpuPath, "possible"))
+	if err != nil {
+		return nil, err
+	}
+
+	return parseCPURange(string(data))
+}
+
+// parseCPURange parses the CPU range string (e.g., "0-3") and returns a list of CPUs.
+func parseCPURange(cpuRange string) ([]int, error) {
```

Comment:
suggestion: Should this return (start ID, end ID, error) rather than ([]IDs, error) ?

*(recreated from https://github.com/neondatabase/autoscaling/pull/1111#discussion_r1804084900)*

Reply:
It is not always possible, input might be for example "0-13,15,17-19". The logic of this particular code is that we always enable/disable in order, but in general the CPU state is shared resource and there are no guarantees that it is kept in order.

Reply:
changed